packages/cli/src/TypedEmitter.ts

```diff
@@ -0,0 +1,35 @@
+import { EventEmitter } from 'node:events';
+import debounce from 'lodash/debounce';
+
+type EventName = string;
+
+type Payloads<L> = {
+	[E in Extract<keyof L, EventName>]: unknown;
+};
+
+type Listener<P> = (payload: P) => void;
+
+export class TypedEmitter<L extends Payloads<L>> extends EventEmitter {
+	protected debounceWait = 300;
+
+	override on<U extends Extract<keyof L, EventName>>(event: U, listener: Listener<L[U]>) {
```

Comment:
Why `U` and not `E`? In general I find all the abbreviations and `Extract` uses might be making this more opaque that it needs to be, esp. as we know in an `EventEmitter` all keys are strings anyway. Here's my take if you think it's useful:

```ts
import { EventEmitter } from 'node:events';
import debounce from 'lodash/debounce';

type Payloads<ListenerMap> = {
	[E in keyof ListenerMap]: unknown;
};

type Listener<Payload> = (payload: Payload) => void;

export class TypedEmitter<ListenerMap extends Payloads<ListenerMap>> extends EventEmitter {
	private debounceWait = 300;

	override on<EventName extends keyof ListenerMap & string>(
		eventName: EventName,
		listener: Listener<ListenerMap[EventName]>,
	) {
		return super.on(eventName, listener);
	}

	override once<EventName extends keyof ListenerMap & string>(
		eventName: EventName,
		listener: Listener<ListenerMap[EventName]>,
	) {
		return super.once(eventName, listener);
	}

	override off<EventName extends keyof ListenerMap & string>(
		eventName: EventName,
		listener: Listener<ListenerMap[EventName]>,
	) {
		return super.off(eventName, listener);
	}

	override emit<EventName extends keyof ListenerMap & string>(
		eventName: EventName,
		payload?: ListenerMap[EventName],
	): boolean {
		return super.emit(eventName, payload);
	}

	protected debouncedEmit = debounce(
		<EventName extends keyof ListenerMap & string>(
			eventName: EventName,
			payload?: ListenerMap[EventName],
		) => super.emit(eventName, payload),
		this.debounceWait,
	);
}
```